### ISSUE TITLE
grad(lambda M, i, j: eigh(M)[1][i, j], argnums=0) returns all NaNs when M has degenerate eigenvalues

### ISSUE THREAD

**Issue by ehermes** [google/jax#4646]
I understand that the eigenvectors of M which correspond to degenerate eigenvalues will have undefined derivatives (because the eigenvectors themselves are poorly defined), but the non-degenerate eigenvectors should have well-defined derivatives. In my application, I need to calculate the derivative of the right-most eigenvector, which should never be degenerate.

A minimal example script:
```python
import jax.numpy as jnp
from jax import grad

def eigh(M, i, j):
    return jnp.linalg.eigh(M)[1][i, j]

deigh = grad(eigh, argnums=0)

M = jnp.array([
    [4., 0., 0.],
    [0., 1., 0.],
    [0., 0., 1.],
], dtype=jnp.float32)

V = jnp.array([[eigh(M, i, j) for j in range(3)] for i in range(3)])
dV = jnp.array([[deigh(M, i, j) for j in range(3)] for i in range(3)])

print(V)
print(dV)
```

Which produces the output:
```
[[0. 0. 1.]
 [1. 0. 0.]
 [0. 1. 0.]]
[[[[nan nan nan]
   [nan nan nan]
   [nan nan nan]]

  [[nan nan nan]
   [nan nan nan]
   [nan nan nan]]

  [[nan nan nan]
   [nan nan nan]
   [nan nan nan]]]


 [[[nan nan nan]
   [nan nan nan]
   [nan nan nan]]

  [[nan nan nan]
   [nan nan nan]
   [nan nan nan]]

  [[nan nan nan]
   [nan nan nan]
   [nan nan nan]]]


 [[[nan nan nan]
   [nan nan nan]
   [nan nan nan]]

  [[nan nan nan]
   [nan nan nan]
   [nan nan nan]]

  [[nan nan nan]
   [nan nan nan]
   [nan nan nan]]]]
```

**Comment by shoyer** [comment]
One way to solve this is by defining a special version of `eigh` with a custom VJP rule. I think the following works:
```python
from functools import partial
from jax import custom_vjp
import jax.numpy as jnp
from jax import lax
import jax.test_util as jtu
import numpy as np

@custom_vjp
def safe_eigh(x):
  return jnp.linalg.eigh(x)

def safe_eigh_fwd(x):
  w, v = safe_eigh(x)
  return (w, v), (w, v)

def safe_eigh_bwd(res, g):
  w, v = res
  wct, vct = g
  deltas = w[..., jnp.newaxis, :] - w[..., :, jnp.newaxis]
  on_diagonal = jnp.eye(w.shape[-1], dtype=bool)
  F = jnp.where(on_diagonal, 0, 1 / jnp.where(on_diagonal, 1, deltas))
  matmul = partial(jnp.matmul, precision=lax.Precision.HIGHEST)
  vT_ct = matmul(v.T.conj(), vct)
  F_vT_vct = jnp.where(vT_ct != 0, F * vT_ct, 0)  # ignore values that would give NaN
  g = matmul(v, matmul(jnp.diag(wct) + F_vT_vct, v.T.conj()))
  g = (g + g.T.conj()) / 2
  return (g,)

safe_eigh.defvjp(safe_eigh_fwd, safe_eigh_bwd)

# unit test
jtu.check_grads(safe_eigh, (np.random.RandomState(0).randn(3, 3),), order=2, modes=['rev'])
```

On your example input, this produces:
```
>>> jax.jacrev(lambda x: safe_eigh(x)[1])(M)  # equivalent to your helper function
DeviceArray([[[[ 0.        , -0.16666667,  0.        ],
               [-0.16666667,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ]],

              [[ 0.        ,  0.        , -0.16666667],
               [ 0.        ,  0.        ,  0.        ],
               [-0.16666667,  0.        ,  0.        ]],

              [[ 0.        ,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ]]],


             [[[ 0.        ,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ]],

              [[        nan,         nan,         nan],
               [        nan,         nan,         nan],
               [        nan,         nan,         nan]],

              [[ 0.        ,  0.16666667,  0.        ],
               [ 0.16666667,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ]]],


             [[[        nan,         nan,         nan],
               [        nan,         nan,         nan],
               [        nan,         nan,         nan]],

              [[ 0.        ,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ],
               [ 0.        ,  0.        ,  0.        ]],

              [[ 0.        ,  0.        ,  0.16666667],
               [ 0.        ,  0.        ,  0.        ],
               [ 0.16666667,  0.        ,  0.        ]]]], dtype=float32)
```

It's a good question whether we can change the default auto-diff rule in JAX to use this. I suspect the answer is "no," because the gradient rule for `eigh` is defined by transposing its JVP rule, but this custom VJP rule is technically non-linear (due to masking with `jnp.where`).

**Comment by shoyer** [comment]
> I am not sufficiently familiar with the inner workings of JAX to fully understand how your solution is meant to work, so apologies if I say something incorrect. I believe this doesn't work for matrices with degenerate eigenvalues:

Well, some of the terms in the Jacobian (those corresponding to degenerate eigenvectors) are going to be NaN no matter way -- the gradient is not well defined mathematically. But the other gradient terms are defined here.

If you exclude the degenerate eigenvectors from the output, then I agree that it should be possible to define a mathematically valid JVP rule. That might be better than using `custom_vjp`. `scipy.linalg.eigh` seems to have some options for partial eigenvetocr problems that could be useful here (see `subset_by_value`): https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eigh.html

**Comment by shoyer** [comment]
It appears that the method for calculating derivatives from https://github.com/google/jax/issues/669#issuecomment-777052841 could solve this problem.
